### PR TITLE
Gpl 263b well extension time

### LINKED ISSUE #497: GPL-263b As Craig I would like the ability to add a pre-extension time to the well for a run as this is required for the sample sheet and I currently have to enter it manually #268
On updating a well, select a Sequencing mode:
If the sequencing mode is CLR, Automation Parameters in the CSV is blank

If the sequencing mode is CCS, a field should appear where they can enter a pre-extension time.
Add pre-extension to a well - number field - not required

Coupled with https://github.com/sanger/traction-service/issues/268

### PR DESCRIPTION
**Pull request by BenTopping**
Closes #497

Changes proposed in this pull request:

*Pre-extension time is now in pacbio well modal
*Pre-extension defaults to 2 for sequencing mode CCS and 0 for CLR
* ...

### CODE REVIEW

File: src/components/pacbio/PacbioWellModal.vue [hunk 1]
```diff
@@ -53,6 +53,19 @@
...
+        <b-form-group
+                      id="preExtensionTime-group"
+                      label="Pre-extension time:"
+                      label-for="preExtensionTime">
+          <b-form-input
+            ref="preExtensionTime"
+            id="preExtensionTime"
+            :value="preExtensionTime"
+            @change="updatePreExtensionTime"
+            placeholder="Pre-extension time">
```
Comment (KatyTaylor) [inline src/components/pacbio/PacbioWellModal.vue:65]: Maybe put `(hours)` to make it clear what unit is required?

File: src/components/pacbio/PacbioWellModal.vue [hunk 2]
```diff
@@ -136,8 +149,19 @@ export default {
       this.mutateWell({ position: this.position, property: 'movie_time', with: movieTime })
     },
     updateSequencingMode(seqMode) {
+      this.preExtensionTimeValue(seqMode)
       this.mutateWell({ position: this.position, property: 'sequencing_mode', with: seqMode })
     },
+    updatePreExtensionTime(preExtensionTime) {
+      this.mutateWell({ position: this.position, property: 'pre_extension_time', with: preExtensionTime })
+    },
+    preExtensionTimeValue(seqMode) {
```
Comment (KatyTaylor) [inline src/components/pacbio/PacbioWellModal.vue:158]: I would call this method 'defaultPreExtensionTime', so it starts with a verb and is a bit more descriptive of what it does

File: src/components/pacbio/PacbioWellModal.vue [hunk 3]
```diff
@@ -149,10 +149,18 @@ export default {
       this.mutateWell({ position: this.position, property: 'movie_time', with: movieTime })
     },
     updateSequencingMode(seqMode) {
+      this.preExtensionTimeValue(seqMode)
```
Comment (harrietc52) [inline src/components/pacbio/PacbioWellModal.vue:152]: Really good idea to put it here 👍

File: tests/unit/components/pacbio/PacbioWell.spec.js [hunk 4]
```diff
@@ -15,6 +15,7 @@ describe('Well.vue', () => {
     storeWell.insert_size = 123
     storeWell.on_plate_loading_concentration = 234
     storeWell.sequencing_mode = "CCS"
+    storeWell.pre_extension_time = "2"
```
Comment (KatyTaylor) [inline tests/unit/components/pacbio/PacbioWell.spec.js:18]: I notice you've got some integers on the other attributes - pre_extension_time should really be an integer too (it is in the database)

File: tests/unit/components/pacbio/PacbioWellModal.spec.js [hunk 5]
```diff
@@ -67,8 +69,11 @@ describe('PacbioRunInfo', () => {
     it('has a table of well libraries', () => {
       expect(wrapper.find('#wellLibraries')).toBeDefined()
     })
+    it('has a Extension Time input', () => {
+      expect(wrapper.find('.insertSize')).toBeDefined()
+    })
```
Comment (KatyTaylor) [inline tests/unit/components/pacbio/PacbioWellModal.spec.js:74]: `Extension Time input` doesn't seem to match `.insertSize` - is this a copy-paste error?

**Comment by KatyTaylor** [comment]
I just pulled your branch in order to test it with mine. It seems to work nicely, except for one thing I tested:
- Edit a well, put 'hello' or some other string in there
- Click Update on the well, then Update on the run
- Go back into the well
- Field displays '0', not 'hello'
- I wasn't shown an error message at any time

We could either 
- validate in the modal that it has to be an integer, or 
- catch the traction-service response, and display an error if it failed (preferably both). 
What do you think? Maybe it's actually traction service's fault, maybe it's not returning an error?

File: tests/unit/components/pacbio/PacbioWell.spec.js [hunk 6]
```diff
@@ -115,8 +116,8 @@ describe('Well.vue', () => {
       expect(ellipse.attributes('class')).toEqual("complete")
     })
 
-    it('will be invalid if sequencing mode is CCS and extension time not present', () => {
-      storeWell.extension_time = ''
+    it('will be valid if sequencing mode is CCS and extension time is present', () => {
+      storeWell.pre_extension_time = '1'
```
Comment (harrietc52) [inline tests/unit/components/pacbio/PacbioWell.spec.js:120]: Setting it to 2 would reflect reality more closeyl

File: tests/unit/components/pacbio/PacbioWell.spec.js [hunk 7]
```diff
@@ -126,12 +127,12 @@ describe('Well.vue', () => {
...
       let ellipse = wrapper.find('ellipse')
-      expect(ellipse.attributes('class')).toEqual("filled")
+      expect(ellipse.attributes('class')).toEqual("complete")
     })
 
-    it('will be valid if sequencing mode is CLR and extension time not present', () => {
+    it('will be valid if sequencing mode is CLR and extension time is empty', () => {
       storeWell.sequencing_mode = 'CLR'
-      storeWell.extension_time = ''
+      storeWell.pre_extension_time = ''
```
Comment (harrietc52) [inline tests/unit/components/pacbio/PacbioWell.spec.js:135]: Would this then be 0?

File: tests/unit/components/pacbio/PacbioWellModal.spec.js [hunk 8]
```diff
@@ -108,9 +108,16 @@ describe('PacbioRunInfo', () => {
       expect(modal.mutateWell).toBeCalledWith({ position: props.position, property: 'sequencing_mode', with: 'CLR' })
     })
 
-    it('updateExtensionTime', () => {
-      modal.updateExtensionTime('1')
-      expect(modal.mutateWell).toBeCalledWith({ position: props.position, property: 'extension_time', with: '1' })
+    it('updatePreExtensionTime', () => {
+      modal.updatePreExtensionTime('1')
+      expect(modal.mutateWell).toBeCalledWith({ position: props.position, property: 'pre_extension_time', with: '1' })
```
Comment (harrietc52) [inline tests/unit/components/pacbio/PacbioWellModal.spec.js:113]: Like Katy said above, making these all integers for consistency would be a good idea. And either 0/2 depending on the mode

File: tests/unit/components/pacbio/PacbioWellModal.spec.js (hunk 5 above)
Comment (BenTopping) [reply tests/unit/components/pacbio/PacbioWellModal.spec.js:74]: yep must be a test i missed false passing

File: tests/unit/components/pacbio/PacbioWell.spec.js [hunk 9]
```diff
@@ -117,7 +117,7 @@ describe('Well.vue', () => {
     })
 
     it('will be valid if sequencing mode is CCS and extension time is present', () => {
-      storeWell.pre_extension_time = '1'
+      storeWell.pre_extension_time = '2'
```
Comment (harrietc52) [inline tests/unit/components/pacbio/PacbioWell.spec.js:120]: Line 135: `storeWell.pre_extension_time = '0'`

**Pull request merged**